packages/flutter_tools/lib/src/build_system/phases.dart

```diff
@@ -0,0 +1,180 @@
+        target.dependencies.add(dependency);
+      }
+    }
+    checkCycles(result);
+    return result;
+  }
+}
+
+// A synthetic target created from a List of [BuildPhase] aggregate targets.
+class _RootTarget extends Target {
+  _RootTarget(this.dependencies, this.definition);
+
+  final BuildDefinition definition;
+
+  @override
+  final List<Target> dependencies;
+
+  @override
+  String get name => definition.name;
+
+  @override
+  Future<void> build(List<File> inputFiles, Environment environment) async { }
+
+  @override
+  List<Source> get inputs => const <Source>[];
+
+  @override
+  List<Source> get outputs => const <Source>[];
+}
+
+// A synthetic target created from the result of a [BuildPhase.plan].
+class _AggregateTarget extends Target {
+  _AggregateTarget(this.dependencies, this.phase);
+
+  final BuildPhase phase;
+
+  @override
+  Future<void> build(List<File> inputFiles, Environment environment) async { }
+
+  @override
+  final List<Target> dependencies;
+
+  @override
+  List<Source> get inputs => const <Source>[];
+
+  @override
+  String get name => phase.name;
+
+  @override
+  List<Source> get outputs => const <Source>[];
+}
+
+/// A Phase is a creator of [Targets].
+///
+/// A phase doesn't have explicit dependencies on individual targets. Instead,
+/// the build system requires all phases in [dependencies] to have run before
+/// this phase.
+///
+/// A phase will always be run, but the targets that constitute it may be
+/// skipped. The [plan] method of a build's phases are always invoked before
+/// any targets are run.
+abstract class BuildPhase {
+  const BuildPhase();
+
+  /// A build phase that consists entirely of static targets.
+  ///
+  /// All of [name], [target], and [dependencies] must not be null.
+  const factory BuildPhase.static({
+    @required String name,
+    @required Target target,
+    @required List<String> dependencies,
+  }) = _StaticBuildPhase;
+
+  /// The name of the build phase.
+  String get name;
+
+  /// The names of the [BuildPhase]s this phase depends on.
+  List<String> get dependencies;
+
+  /// Generate one or more targets for this build phase.
+  Future<List<Target>> plan(Environment environment);
+}
+
+class _StaticBuildPhase implements BuildPhase {
+  const _StaticBuildPhase({
+    @required this.name,
+    @required this.target,
+    @required this.dependencies
+  }) : assert(name != null),
+       assert(target != null),
+       assert(dependencies != null);
+
+  @override
+  final List<String> dependencies;
+
+  @override
+  final String name;
+
+  final Target target;
+
+  @override
+  Future<List<Target>> plan(Environment environment) async {
+    return <Target>[target];
+  }
+}
+
+/// Check if there are any dependency cycles in the target.
+///
+/// Throws a [CycleException] if one is encountered.
+void checkCycles(Target initial) {
+  void checkInternal(Target target, Set<String> visited, Set<String> stack) {
+    if (stack.contains(target.name)) {
```

Comment:
Consider the following two graphs:

I'm not sure what you mean. The stack is the set of targets we've traversed to get to this point. For Example (valid graph):

```
a -> b -> d
     c -> d
```
When we're looking at `d` it will either contain `{a, b, d}` or `{a, c, d}`. This is due to the pop below on L177. Something like:
```
{ } visit a
{ a } visit b
{ a, b } visit d
{ a, b, d } pop d
{ a, b } pop b
{ a, } visit c
{ a, c } visit d
{ a, c, d } 
```

Reply:
Sure. I was wondering more about what this target does. Is this something like a reusable target, e.g. clone file?